localstack-core/localstack/services/apigateway/next_gen/execute_api/router.py

```diff
@@ -134,14 +142,32 @@ def register_routes(self) -> None:
                 endpoint=self.handler,
                 strict_slashes=True,
             ),
-            # add the localstack-specific _user_request_ routes
+            # add the deprecated localstack-specific _user_request_ routes
             self.router.add(
                 path="/restapis/<api_id>/<stage>/_user_request_",
-                endpoint=self.handler,
+                endpoint=deprecated_route_endpoint,
                 defaults={"path": ""},
             ),
             self.router.add(
                 path="/restapis/<api_id>/<stage>/_user_request_/<greedy_path:path>",
+                endpoint=deprecated_route_endpoint,
+                strict_slashes=True,
+            ),
+            # add the localstack-specific so-called "path-style" routes when DNS resolving is not possible
+            self.router.add(
+                path=f"{self.EXECUTE_API_INTERNAL_PATH}/<api_id>/",
+                endpoint=self.handler,
+                defaults={"path": "", "stage": None},
+                strict_slashes=True,
+            ),
+            self.router.add(
+                path=f"{self.EXECUTE_API_INTERNAL_PATH}/<api_id>/<stage>/",
+                endpoint=self.handler,
+                defaults={"path": ""},
+                strict_slashes=False,
+            ),
+            self.router.add(
+                path=f"{self.EXECUTE_API_INTERNAL_PATH}/<api_id>/<stage>/<greedy_path:path>",
                 endpoint=self.handler,
                 strict_slashes=True,
             ),
```

Comment:
A better strategy might be to match only `<api_id>` and `<api_id>/<greedy_path:path>` and have to do a bit more work when routing the request. It is added work on the rest api but would help us support HTTP api with less mess

Reply:
We had a lengthy discussion about this, and the TLDR; is this:

We're using the router to do as much parsing as it can, and if we were to remove the stage from the path, we would still need to manually parse it from every request by checking the first "resource" of the path and assigning it to the stage name.

For HTTP API, you can define a `$default` stage, so if you do not specify a stage name, it will fallback to the default. That leads to the registered routes being somewhat misleading, because if you target an HTTP API with a $default stage configured with the following path: `test/test-path`, then in our edge router endpoint, `test` will be assigned to the stage value.
The next check is to verify if that `test` is in the existing `stages`, and if not, assign the stage value to `$default`.

We don't use the `<greedy_path:path>` value in the rest of the invocation, as it is URL-decoded, and we need the raw value. In the rest of the logic, we first use `RAW_URI` which contains the entire path of the request, and we do manipulation to remove the stage name if the request is prefixed with it. The `stage` parameter here is only to help us get that first "resource" of the path, it being rightly the stage name or not is not really important, we do not know yet if the request is proper or not. 

We've also decided to model the edge router rules to the REST API format, as `stage` is mandatory. It will be a bit of a workaround to support all kind of API with the same routes, but if properly documented, should be alright. 

This Edge router endpoint can be seen as a "first pass" routing, trying to catch all API Gateway request and doing as much work as possible, and then it's up to us to properly route the request to the right API Gateway Resource/Route of the API.